06_puppy/sirigithub/syncStore.go

```diff
@@ -0,0 +1,63 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+)
+
+type SyncStore struct {
+	sync.Map
+	sync.Mutex
+}
+
+func NewSyncStore() *SyncStore {
+	return &SyncStore{}
+}
+
+/** CreatePuppy takes a Puppy with a user provided ID, and creates a new Puppy if
+the ID does not already exist in the store.
+**/
+func (m *SyncStore) CreatePuppy(puppy *Puppy) error {
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.Load(puppy.ID); ok {
+		return fmt.Errorf("puppy with id %d already exists", puppy.ID)
+	}
+	m.Store(puppy.ID, *puppy)
+	return nil
+}
+
+/** UpdatePuppy overrides an existing puppy in the store,
+returns an error if id is not found or does not match the puppy ID**/
+func (m *SyncStore) UpdatePuppy(id int, p *Puppy) error {
+	m.Lock()
+	defer m.Unlock()
+	if _, exists := m.Load(id); !exists {
+		return fmt.Errorf("puppy %d to update does not exist in the map", id)
+	}
+	if id != p.ID {
+		return fmt.Errorf("puppy Id %d to update does not match %d", id, p.ID)
+	}
+	m.Store(id, p)
+	return nil
+}
+
+/** DeletePuppy deletes an existing puppy from the store **/
+func (m *SyncStore) DeletePuppy(id int) (bool, error) {
+	m.Lock()
+	defer m.Unlock()
+	if _, exists := m.Load(id); exists {
+		m.Delete(id)
+		return true, nil
```

Comment:
If you return an error, it seems redundant to still return bool

Reply:
Thanks for the review @hsy3418. Agree and updates implemented.